src/System.Console/src/System/ConsolePal.Unix.cs

```diff
@@ -356,202 +410,228 @@ private static unsafe void GetCursorPosition(out int left, out int top)
-                // the timeout on the very first request to 15 seconds, to account for potential latency
-                // before we know if we will receive a response.
-                Interop.Sys.InitializeConsoleBeforeRead(minChars: (byte)(s_everReceivedCursorPositionResponse ? 1 : 0), decisecondsTimeout: (byte)(s_firstCursorPositionRequest ? 100 : 10));
-                try
+                int cursorVersion;
+                lock (s_cursorGate)
                 {
-                    // Write out the cursor position report request.
-                    Debug.Assert(!string.IsNullOrEmpty(TerminalFormatStrings.CursorPositionReport));
-                    WriteStdoutAnsiString(TerminalFormatStrings.CursorPositionReport);
-
-                    // Read the cursor position report (CPR), of the form \ESC[row;colR. This is not
-                    // as easy it it sounds.  Prior to the CPR having been supplied to stdin, other
-                    // user input could have come in and be available to read first from stdin.  Plus,
-                    // that user input could include escape sequences, and those escape sequences could
-                    // have a prefix very similar to that of the CPR (e.g. other escape sequences start
-                    // with \ESC + '['.  It's also possible that some terminal implementations may not
-                    // write the CPR to stdin atomically, such that the CPR could have other user input
-                    // in the middle of it, and that user input could have escape sequences!  Handling
-                    // that last case is very challenging, and rare, so we don't try, but we do need to
-                    // handle the rest.  The min bar here is doing something reasonable, which may include
-                    // giving up and just returning default top and left values.
-
-                    // Consume from stdin until we find all of the key markers for the CPR:
-                    // \ESC, '[', ';', and 'R'.  For everything before the \ESC, it's definitely
-                    // not part of the CPR sequence, so we just immediately move any such bytes
-                    // over to the StdInReader's extra buffer.  From there until the end, we buffer
-                    // everything into readBytes for subsequent parsing.
-                    const byte Esc = 0x1B;
-                    StdInReader r = StdInReader.Inner;
-                    int escPos, bracketPos, semiPos, rPos;
-                    if (!AppendToStdInReaderUntil(Esc, r, readBytes, ref readBytesPos, out escPos) ||
-                        !BufferUntil((byte)'[', r, ref readBytes, ref readBytesPos, out bracketPos) ||
-                        !BufferUntil((byte)';', r, ref readBytes, ref readBytesPos, out semiPos) ||
-                        !BufferUntil((byte)'R', r, ref readBytes, ref readBytesPos, out rPos))
+                    if (HasCachedCursorPosition())
```

Comment:
It looks like everywhere we call this we then also want the left and top... should we make it instead be `TryGetCachedCursorPosition(out int left, out int top)` or something like that?